src/Std/System/Uri.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+def hexDigitToNat? (c : Char) : Option Nat :=
+  if '0' ≤ c ∧ c ≤ '9' then some (c.toNat - '0'.toNat)
+  else if 'a' ≤ c ∧ c ≤ 'f' then some (c.toNat - 'a'.toNat + 10)
+  else if 'A' ≤ c ∧ c ≤ 'F' then some (c.toNat - 'A'.toNat + 10)
+  else none
+
+/-- Decode %HH escapings in the given string. Note that sometimes a consecutive
+sequence of multiple escapings can represet a utf-8 encoded sequence for
+a single unicode code point and these will also be decoded correctly. -/
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let len := uri.utf8ByteSize
+  let mut i := uri.mkIterator
+  while !i.atEnd do
+    let pos := i.2.byteIdx
+    let c := i.curr
+    if c == '%' && pos + 1 < len then
+      let h1 : Char := i.next.curr
+      if h1 == '%' then
+        -- this is an escaped '%%' which should become a single percent.
+        decoded := decoded.push 37 -- %
+        i := i.next.next
+      else if pos + 2 < len then
+        -- should have %HH where HH are hex digits, if they are not
+        -- valid hex digits then just repeat whatever sequence of chars
+        -- we found here.
+        let h2 := i.next.next.curr
+        let d1 := hexDigitToNat? h1
+        let d2 := hexDigitToNat? h2
+        decoded := if let (some hd1, some hd2) := (d1, d2) then
+          decoded.push (hd1 * 16 + hd2).toUInt8
+        else
+          decoded.push 37 ++ h1.toString.toUTF8 ++ h2.toString.toUTF8
+        i := i.nextn 3
+      else
+        decoded := decoded.push 37 ++ h1.toString.toUTF8
+        i := i.next.next
+
+    else
+      decoded := decoded.append c.toString.toUTF8
+      i := i.next
+   return String.fromUTF8Unchecked decoded
+
+/- https://datatracker.ietf.org/doc/html/rfc3986/#page-12 -/
+def rfc3986ReservedChars : List Char := [ ';', ':', '?', '#', '[', ']', '@', '&', '=', '+', '$', ',', '!', '\'', '(', ')', '*', '%' ]
+
+def uriEscapeAsciiChar (c : Char) : String :=
+  if (rfc3986ReservedChars.contains c) then
+    "%" ++ smallCharToHex c
+  else if ((Char.toNat c) < 127) then
+    c.toString
+  else
+    c.toString.toUTF8.foldl (fun s c => s ++ "%" ++ (smallCharToHex (Char.ofNat c.toNat))) ""
+  where
+  smallCharToHex (c : Char) : String :=
+    let n  := Char.toNat c;
+    let d2 := n / 16;
+    let d1 := n % 16;
+    hexDigitRepr d2 ++ hexDigitRepr d1
+end UriEscape
+
+/-- Convert the given file name to a "file:///encodedpath" Uri
+where the encoded path may contain %xx escaping when needed. -/
+def toFileUri (s : String) : String := Id.run do
```

Comment:
The FilePath helpers are handy and are being used by the LSP Utils instead of LSP Utils defining it's own local FilePath version, plus the FilePath case has to be careful to use "normalized" strings in `pathToUri` and if we have `pathToUri` we have to have the counter part `fileUriToPath?`

Reply:
I should have been clearer: I think we should have only `pathToUri` and `fileUriToPath?`. If a parameter is representing a file path, it should always be a `FilePath` in public APIs.

Reply:
Ah, I see, done.